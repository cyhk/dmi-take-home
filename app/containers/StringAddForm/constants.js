export const ADD_STRING = 'ADD_STRING';
export const ADD_STRING_SUCCESS = 'ADD_STRING_SUCCESS';
export const ADD_STRING_FAILURE = 'ADD_STRING_FAILURE';
