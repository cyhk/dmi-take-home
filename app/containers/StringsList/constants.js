export const GET_STRINGS = 'GET_STRINGS';
export const GET_STRINGS_SUCCESS = 'GET_STRINGS_SUCCESS';
export const GET_STRINGS_FAILURE = 'GET_STRINGS_FAILURE';
