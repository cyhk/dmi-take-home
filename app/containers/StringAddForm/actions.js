import { ADD_STRING_SUCCESS, ADD_STRING_FAILURE } from './constants';

export function stringAdded(strings) {
  return { type: ADD_STRING_SUCCESS, strings };
}

export function stringAddingError(error) {
  return { type: ADD_STRING_FAILURE, error };
}
