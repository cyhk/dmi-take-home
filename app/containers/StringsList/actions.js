import { GET_STRINGS_SUCCESS, GET_STRINGS_FAILURE } from './constants';

/**
 * Passes along strings array
 */
export function stringsLoaded(strings) {
  return { type: GET_STRINGS_SUCCESS, strings };
}

/**
 * Passes along error
 */
export function stringsLoadingError(error) {
  return { type: GET_STRINGS_FAILURE, error };
}
