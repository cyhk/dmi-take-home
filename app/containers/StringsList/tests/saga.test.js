/**
 * Tests for StringsList sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { GET_STRINGS } from '../constants';
import { stringsLoaded, stringsLoadingError } from '../actions';

import getStringsWatcherSaga, { getStringsWorkerSaga } from '../saga';

const string = 'test';

/* eslint-disable redux-saga/yield-effects */
describe('getStringsWorkerSaga', () => {
  let getStringsWorkerGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getStringsWorkerGenerator = getStringsWorkerSaga();

    const callDescriptor = getStringsWorkerGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the stringAdded action if it posts the data successfully', () => {
    const stringsMock = [string];
    const responseMock = {
      data: { strings: stringsMock },
    };
    const putDescriptor = getStringsWorkerGenerator.next(responseMock).value;
    expect(putDescriptor).toEqual(put(stringsLoaded(stringsMock)));
  });

  it('should call the stringAddingError action if the response errors', () => {
    const response = new Error('TestError');
    const putDescriptor = getStringsWorkerGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(stringsLoadingError(response)));
  });
});

describe('getStringsWatcherSaga', () => {
  const getStringsWatcher = getStringsWatcherSaga();

  it('should start task to watch for ADD_STRING action', () => {
    const takeLatestDescriptor = getStringsWatcher.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(GET_STRINGS, getStringsWorkerSaga),
    );
  });
});
