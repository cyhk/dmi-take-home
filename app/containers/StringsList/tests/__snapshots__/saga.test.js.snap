// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStringsWorkerSaga should call the stringAddingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getStringsWorkerSaga should dispatch the stringAdded action if it posts the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
